# Uncommitted local edits in source/ relative to its SVN base (file:///D:/Repositories/FECells, trunk)
# Generated by comparing each file against .svn/pristine/

--- biology.cpp (trunk r517)
+++ biology.cpp (working copy)
@@ -23,6 +23,8 @@
 #include <libBioModel/GaussianRNG.h>
 #include <sbml/Model.h>
 
+#define MIN_FBEHAVIOR_PERSISTENCE 5
+
 using namespace std;
 
 Biology::Biology()
@@ -165,7 +167,7 @@
 	D[2]  = nw_common->getModel()->getParameter("D_collagen")->getValue();   // 1.0E-9; cm^2/s, 2. collagen
 	D[3]  = nw_common->getModel()->getParameter("D_PDGF")->getValue();    // 2.78E-8; cm^2/s, 3. PDGF
 	D[4]  = nw_common->getModel()->getParameter("D_KGF")->getValue();  // 9.5E-6; cm^2/s, 4.KGF(FGF7)
-	D[5]  = nw_common->getModel()->getParameter("D_EGF")->getValue(); // 5.18E-7; cm^2/s, 5. EGF (Thorne et al. 2004 J Neurophysiol 92:3471-3481.)
+	D[5]  = nw_common->getModel()->getParameter("D_IL1")->getValue(); // 5.18E-7; cm^2/s, 5. EGF (Thorne et al. 2004 J Neurophysiol 92:3471-3481.)
 	D[6]  = nw_common->getModel()->getParameter("D_TGFb1")->getValue(); // 2.94E-7; cm^2/s, 6. TGF-beta1
 	D[7]  = nw_common->getModel()->getParameter("D_Procollagen")->getValue(); // cm^2/s, 7. procollagen
 	D[8]  = nw_common->getModel()->getParameter("D_MMP")->getValue(); // 8.0E-9; cm^2/s, 9. MMP-1 (Saffarian et al. 2004 Science 306:108-111.)
@@ -699,10 +701,10 @@
 	adhesion_pair_array[3][4] = 0;   // k-FC      3-4
 	adhesion_pair_array[4][4] = 2;   // FC-FC     4-4
 	adhesion_pair_array[0][6] = 2.3; // ECM-MP    0-6
-	adhesion_pair_array[1][6] = 2;   // BM-MP     1-6
+	adhesion_pair_array[1][6] = 0;   // BM-MP     1-6
 	adhesion_pair_array[2][6] = 0;   // f-MP      2-6
 	adhesion_pair_array[3][6] = 0;   // k-MP      3-6
-	adhesion_pair_array[4][6] = 2;   // FC-MP     4-6
+	adhesion_pair_array[4][6] = 0;   // FC-MP     4-6
 	for (int i=1;i<20;i++)
 	{
 		for (int j=0;j<i;j++)
@@ -772,7 +774,7 @@
 int dbReader::cell_setupbiobehaviors(int Time)
 {
 	// INTERFACE: MAIN ENTRY
-	int MODE = 1; // SETUP A MODELING MODE, 0: INITIALIZATION; and 1: MODELING; and 2: DELETING AND RELAXING CELLS.
+	int MODE = 0; // SETUP A MODELING MODE, 0: INITIALIZATION; and 1: MODELING; and 2: DELETING AND RELAXING CELLS.
 	cell *pcell0 = 0, *pcell1 = 0;
 	point3D *s1 = 0;
 	point3D *t1 = 0;
@@ -947,6 +949,7 @@
 					pcell0 = cellList[i];
 				}
 
+				///// NEIGHBOR? /////
 				commedgelen = 0.0;
 				s1 = cellList[i]->get_pairs()->at(j)->MP()->at(0)->S1();
 				t1 = cellList[i]->get_pairs()->at(j)->MP()->at(0)->T1();
@@ -976,14 +979,33 @@
 					}
 				}
 				if (s1==t1) {commedgelen = 0.1;}
-				if (commedgelen > 0)
+				int idx_i = 0;
+				int idx_j = 0;
+				if (commedgelen > 0) ///// NEIGHBOR! /////
 				{
-					ia.push_back (cellid2idx[pcell1->id()]);
-					ja.push_back (cellid2idx[i]);
+					double distsq = 0;
+					distsq =  (pcell0->get_center(0) - pcell1->get_center(0)) * (pcell0->get_center(0) - pcell1->get_center(0));
+					distsq += (pcell0->get_center(1) - pcell1->get_center(1)) * (pcell0->get_center(1) - pcell1->get_center(1));
+					idx_i = cellid2idx[pcell1->id()];
+					idx_j = cellid2idx[i];
+					ia.push_back (idx_i);
+					ja.push_back (idx_j);
 					for(int k=0; k<biology->getNumRealSpecies(); k++)
 					{
-						a[k].push_back (biology->getD(k) * biology->getD0() * commedgelen);
-						vii[k] += biology->getD(k) * biology->getD0() * commedgelen;
+						if (pcell1->get_cell_type() == 3 && (k == 0 || k == 1 || k == 2 || k == 7 || k == 8))
+						{
+							a[k].push_back(0.0);
+						}
+						else
+						{
+							//a[k].push_back (biology->getD(k) * biology->getD0() * commedgelen);
+							//vii[k] += biology->getD(k) * biology->getD0() * commedgelen;
+							//double value = biology->getD(k) * biology->getD0() * commedgelen;
+							//double value = biology->getD(k) * biology->getD0() * commedgelen / distsq;
+							double value = biology->getD(k) * biology->getD0() / distsq;  // CORRECT
+							a[k].push_back (value);
+							vii[k] += value;
+						}
 					}
 				}
 			}
@@ -1003,7 +1025,8 @@
 	double err = 0;
 	for (int i=0; i<biology->getNumRealSpecies(); i++)
 	{
-		err = Euler (&ia, &ja, &a[i], &v[i], biology->getDeltaT());
+		//err = Euler (&ia, &ja, &a[i], &v[i], biology->getDeltaT());
+		err = RK4 (&ia, &ja, &a[i], &v[i], biology->getDeltaT());
 		cout << "       Diffusion Solved for " << biology->getSpeciesName(i) << ", err=" << err << endl;
 	}
 	cout << "     All Diffusions Done." << endl;
@@ -1020,6 +1043,25 @@
 		}
 	}
 	cout << "  4. Updating Cell States by Diffusion, Done." << endl;
+
+	// IF FIBROBLASTS SYNTHESIZED ENOUGH PROCOLLAGEN, THEN THEY GENERATE NEW ECM ELEMENTS.
+	// WILL FINISH IT LATER!!! CORRELATE ECM ELEMENT WITH COLLAGEN PRODUCTION
+	/*for (int i=0; i<cn; i++)
+	{
+		if (cellList[i]->get_cell_type() == 2) // IF FIBROBLAST, 
+		{
+			int procol = cellList[i]->get_cell_state(7);
+			if (procol > 200) // IF FIBROBLAST AND IT HAS ENOUGH PROCOLLAGEN
+			{
+				ra = (int)cellList[i]->shape_based_division_angle();
+				cell_division(cellList[i], Time, cellList[i]->get_cell_type(), 0, ra);
+				cellList[i]->center_refresh();
+				cellList[cellList.size()-1]->center_refresh();
+				cellList[cellList.size()-1]->set_cell_state(7, 100);
+				cellList[i]->set_cell_state(7, procol - 100);
+			}
+		}
+	}*/
 
 	// STEP 5: Setup Cell Behaviors - A 4-way stochastic switch
 	// (0) TBD; (1) Do nothing; (2) Growth/Division; (3) Apoptosis; (4) Migration.
@@ -1052,7 +1094,7 @@
 			biology->setValueProbApoptosis (biology->getValueProbApoptosis()/sum);
 			double tau1 = ((double)rand()+1.0)/((unsigned long)RAND_MAX+1);
 			// TEST
-			if (cellList[i]->get_cell_type () == 2){
+			/*if (cellList[i]->get_cell_type () == 2){
 			// || cellList[i]->get_cell_type () == 3) {
 			//if (i == 7105) {
 			cout << "  >> Cell behavior prob:";
@@ -1061,7 +1103,7 @@
 			printf(" GrowthDiv=%0.2f", biology->getValueProbGrowthDiv()); 
 			printf(" Migration=%0.2f", biology->getValueProbMigration()); 
 			printf(" Apoptosis=%0.2f", biology->getValueProbApoptosis()); 
-			cout << endl;}/**/
+			cout << endl;}*/
 			//
 			if (tau1 >= 0 && tau1 < biology->getValueProbGrowthDiv())
 			{
@@ -1169,9 +1211,34 @@
 					cout<<"  -> ";
 					// All divisions are self-renewal type
 					cell_division(cellList[i], Time, cellList[i]->get_cell_type(), cellList[i]->get_cell_type(), ra);
+					cellList[i]->center_refresh();
+					cellList[cellList.size()-1]->center_refresh();
+					for (int j=0; j<biology->getNumRealSpecies(); j++)
+					{
+						cellList[cellList.size()-1]->set_cell_state(j, (int)(cellList[i]->get_cell_state(j)/2.0));
+						cellList[i]->set_cell_state(j, (int)(cellList[i]->get_cell_state(j)/2.0));
+					}
 				}
 			}
 		}
+
+		//
+		if (cellList[i]->get_cell_type() == 0)
+		{
+			if (cellList[i]->get_area() >= AR)
+			{
+				//cout << ">>>>TEST<<<< COL_DIV " << cellList[i]->get_area() << endl;
+				ra = (int)cellList[i]->shape_based_division_angle();
+				cell_division(cellList[i], Time, cellList[i]->get_cell_type(), cellList[i]->get_cell_type(), ra);
+				cellList[i]->center_refresh();
+				cellList[cellList.size()-1]->center_refresh();
+				for (int j=0; j<biology->getNumRealSpecies(); j++)
+				{
+					cellList[cellList.size()-1]->set_cell_state(j, (int)(cellList[i]->get_cell_state(j)/2.0));
+					cellList[i]->set_cell_state(j, (int)(cellList[i]->get_cell_state(j)/2.0));
+				}
+			}
+		}
 	}
 	cout << "  7. Doing Cell Divisions, Done." << endl;
 
@@ -1201,6 +1268,16 @@
 				}
 			}
 		}
+
+		// 
+		if (cellList[i]->get_cell_type() == 0)
+		{
+			if (cellList[i]->get_cell_state(2)/1000 > cellList[i]->get_area()/AR0) // ECM VOLUME SHOULD REFLECT ITS MASS OF COLLAGEN.
+			{
+				//cout << ">>>>TEST<<<< COL " << cellList[i]->get_cell_state(2) << " AREA " << cellList[i]->get_area() << endl;
+				cellList[i]->set_GR(AR0 * (cellList[i]->get_cell_state(2)/1000 - cellList[i]->get_area()/AR0));
+			}
+		}/**/
 	}
 	cout << "  8. Doing Cell Growths, Done." << endl;
 
@@ -1329,7 +1406,8 @@
 					cellList[i]->set_migrate(1);
 					cellList[i]->set_migrate_relax(1);
 					//double force = max(0.0, (maxcytokine - (double)cellList[i]->get_cell_state(MIGCYTOKINEID))); // Variable force depending on signal strength
-					double force = GAR*5; // Constant initial force
+					//double force = GAR*5; // Constant initial force
+					double force = GAR*2.5; // Constant initial force
 					cell *Cod = cell_migrate_angle(cellList[i], angle);
 					if (cdist_maxcytokine == 0.0 && Cod!=NULL && Cod==cellList[maxcytokine_id])
 					{
@@ -1377,13 +1455,14 @@
 					}
 					cellList[i]->set_migrate_angle(angle_ow);
 					cell_migrate_response_1(cellList[i], angle_ow, &kct,  2*force); // MIGRATION SPEED
-					cellList[i]->set_GR(1.236 * force);
+					//cellList[i]->set_GR(1.236 * force);
+					cellList[i]->set_GR(1.8*force);
 
-					// Deposite new ECMs
+					// Fibroblasts deposite new ECMs
 					ra = (int)angle_ow + 90;
 					if (ra < 0) { ra = ra + 360; }
 					else if (ra > 360) { ra = ra - 360; }
-					if (cellList[i]->get_area() > AR)
+					if (cellList[i]->get_area() > AR)// && cellList[i]->get_cell_state(7) > 200)
 					{
 						cell_division(cellList[i], Time, cellList[i]->get_cell_type(), cellList[i]->get_cell_type(), ra);
 						cellList[i]->center_refresh();
@@ -1397,15 +1476,20 @@
 						double cos_an = cos(an*PI/180.0);
 						double sin_an = sin(an*PI/180.0);
 						double plot_an = cos_angle_ow*cos_an + sin_angle_ow*sin_an;*/
+						double procoldensity = 0.0;
 						if      (abs(an-angle_ow) < 90 || abs(an-angle_ow) > 270)
-						{ 
+						{
+							procoldensity = cellList[i]->get_cell_state (7) / 2.0;
 							cellList[cellList.size()-1]->set_cell_type(0);
-							cellList[cellList.size()-1]->set_cell_state(2, 100); // Deposit new collagen
+							cellList[cellList.size()-1]->set_cell_state(7, (int)procoldensity); // Deposit new collagen
+							cellList[i]->set_cell_state(7, (int)procoldensity);
 						}
 						else if (abs(an-angle_ow) >= 90 && abs(an-angle_ow)<=270)
-						{ 
+						{
+							procoldensity = cellList[cellList.size()-1]->get_cell_state (7) / 2.0;
 							cellList[i]->set_cell_type(0);
-							cellList[i]->set_cell_state(2, 100); // Deposit new collagen
+							cellList[i]->set_cell_state(7, (int)procoldensity); // Deposit new collagen
+							cellList[cellList.size()-1]->set_cell_state(7, (int)procoldensity);
 							cellList[i]->setup_Mark_number(0);
 							cellList[cellList.size()-1]->setup_Mark_number(1);
 							cellList[cellList.size()-1]->set_migrate_angle(angle_ow);
@@ -1422,7 +1506,7 @@
 					}
 					//// END deposition of new ECMs
 
-					if (cellList[i]->get_bhsteps(4) > 5) // At least migrate for 5 steps.
+					if (cellList[i]->get_bhsteps(4) > MIN_FBEHAVIOR_PERSISTENCE) // At least migrate for 30 steps.
 					{
 						cellList[i]->set_behavior(0);
 						cellList[i]->set_bhsteps(0,0);
@@ -1436,10 +1520,12 @@
 		}
 		else if (cellList[i]->get_cell_type () == 3) // KERATINOCYTES MIGRATION
 		{
+			/* NEW K MIGRATION MODEL
 			// NEW K MIGRATION USING MECHANICAL SIGNALS
 			// MIGRATION SPEED COMES FROM CYTOKINE, BUT DIRECTION COMES FROM MECHANICAL SIGNALS
 			if (cellList[i]->get_behavior() == 4)
 			{
+
 				// Cell MIGRATION
 				cellList[i]->set_GR(0.0);
 				int MIGCYTOKINEID = 4;  // 4: KGF
@@ -1637,9 +1723,9 @@
 						cellList[i]->set_bhsteps(4, cellList[i]->get_bhsteps(4)+1);
 					}
 				}
-			}
+			}*/
 
-			/* OLD K MIGRATION WITHOUT MECHANICAL SIGNALS
+			/* OLD K MIGRATION WITHOUT MECHANICAL SIGNALS */
 			if (cellList[i]->get_behavior() == 4)
 			{
 				cellList[i]->set_GR(0.0);
@@ -1719,15 +1805,19 @@
 					// Find neighboring ECM and Clot cells.
 					int celltype0 = cellList[cell_angles[j][0]]->get_cell_type();
 					int celltype1 = cellList[cell_angles[j+1][0]]->get_cell_type();
-					if ((celltype0 <= 2 || celltype0 == 6 ) && (celltype1 == 3 || celltype1 == 4))
+					if ((celltype0 <= 2) && (celltype1 == 3 || celltype1 == 4 || celltype1 == 6))
 					{
-						if (celltype1 == 4)  // IF NEXT NEIGHBOR IS CLOT, MIGRATE TO THAT DIRECTION
+						if (celltype1 == 4 || celltype1 == 6)  // IF NEXT NEIGHBOR IS CLOT, MIGRATE TO THAT DIRECTION
 						{
 							migangle = cell_angles[j+1][1];
 							maxcytokine = cellList[cell_angles[j+1][0]]->get_cell_state(MIGCYTOKINEID);
 							maxcytokine_id = cell_angles[j+1][0];
+							if (celltype1 == 6)
+							{
+								cellList[cell_angles[j+1][0]]->set_cell_type(4);
+							}
 						}
-						else  // ELSE, LET'S COMPARE EGF GRADIENT
+						else  // ELSE, LET'S COMPARE KGF GRADIENT
 						{
 							if (maxcytokine_id < 0)
 							{
@@ -1749,13 +1839,17 @@
 							}
 						}
 					}
-					else if ((celltype1 <= 2 || celltype1 == 6 ) && (celltype0 == 3 || celltype0 == 4))
+					else if ((celltype1 <= 2) && (celltype0 == 3 || celltype0 == 4 || celltype0 == 6))
 					{
-						if (cellList[cell_angles[j][0]]->get_cell_type() == 4)  // IF NEXT NEIGHBOR IS CLOT, MIGRATE TO THAT DIRECTION
+						if (celltype0 == 4)  // IF NEXT NEIGHBOR IS CLOT, MIGRATE TO THAT DIRECTION
 						{
 							migangle = cell_angles[j][1];
 							maxcytokine = cellList[cell_angles[j][0]]->get_cell_state(MIGCYTOKINEID);
 							maxcytokine_id = cell_angles[j][0];
+							if (celltype0 == 6)
+							{
+								cellList[cell_angles[j][0]]->set_cell_type(4);
+							}
 						}
 						else
 						{
@@ -1831,12 +1925,14 @@
 
 					// DO NOT PUSH FRONT KERATINOCYTE TOO HARD
 					double ffactor = 1.0;
-					if (Cod!=NULL && Cod->get_cell_type() == 3) 
+					if (Cod!=NULL && Cod->get_cell_type() == 3)
 					{
 						ffactor = 0.2;
 					}
 					//double force = ffactor * max(0.0, min(GAR, (double)cellList[i]->get_cell_state(MIGCYTOKINEID)));
-					double force = ffactor * min(2.0*GAR, max(0.0, (maxcytokine - (double)cellList[i]->get_cell_state(MIGCYTOKINEID))));
+					//double force = ffactor * min(2.0*GAR, max(0.0, (maxcytokine - (double)cellList[i]->get_cell_state(MIGCYTOKINEID))));
+					//cout << "    >>>> KMigForceFactor: " << biology->CalcValueMigForce(cellList[i]) << endl;
+					double force = ffactor * biology->CalcValueMigForce(cellList[i]) * GAR * 5;
 					cell_migrate_response_1(cellList[i], migangle, &kct, force);
 					//cellList[i]->set_GR(0.6 * force);
 					cellList[i]->set_GR(force);
@@ -1850,9 +1946,15 @@
 						cell_division(cellList[i], Time, cellList[i]->get_cell_type(), cellList[i]->get_cell_type(), ra);
 						cellList[i]->center_refresh();
 						cellList[cellList.size()-1]->center_refresh();
+						for (int j=0; j<biology->getNumRealSpecies(); j++)
+						{
+							cellList[cellList.size()-1]->set_cell_state(j, (int)(cellList[i]->get_cell_state(j)/2.0));
+							cellList[i]->set_cell_state(j, (int)(cellList[i]->get_cell_state(j)/2.0));
+						}
+
 					}
 					//// END ////
-					if (cellList[i]->get_bhsteps(4) > 10) // At least migrate for 5 steps.
+					if (cellList[i]->get_bhsteps(4) > 5) // At least migrate for 5 steps.
 					{
 						cellList[i]->set_behavior(0);
 						cellList[i]->set_bhsteps(0,0);
@@ -1863,7 +1965,7 @@
 					}
 				}
 			}
-			*/ // OLD MIGRATION HAVE NO MECHANICAL SIGNALS
+			// OLD MIGRATION HAVE NO MECHANICAL SIGNALS
 		}
 	}
 	cout << "  9. Doing Cell Migrations, Done." << endl;
@@ -1875,10 +1977,11 @@
 
 		if (cellList[i]->get_cell_type() == 4)
 		{
-			if (cellList[i]->get_cell_state(1) > cellList[i]->get_cell_state(0))
+			//if (cellList[i]->get_cell_state(2) > 100 && cellList[i]->get_cell_state(1) < 10)
+			if (cellList[i]->get_cell_state(2) > 2 * cellList[i]->get_cell_state(1))
 			{
 				// If clot has more collagen than clot, then change cell type to ECM
-				//cellList[i]->set_cell_type(0);
+				cellList[i]->set_cell_type(0);
 			}
 		}
 
@@ -1890,7 +1993,7 @@
 			{
 				// Cell Growth Rate zero
 				cellList[i]->set_GR(0.0);
-				if (cellList[i]->get_bhsteps(1) >= 0) // Stay dormant for 5 steps.
+				if (cellList[i]->get_bhsteps(1) >= MIN_FBEHAVIOR_PERSISTENCE) // Stay dormant for 5 steps.
 				{
 					cellList[i]->set_behavior(0);
 					cellList[i]->set_bhsteps(0,0);
--- cellproject.cpp (trunk r495)
+++ cellproject.cpp (working copy)
@@ -31,7 +31,7 @@
 	show_text = false;
 	show_direction = true;
 	show_force = false;
-	diffusion_number = 1;
+	diffusion_number = 0;
 	number = 0;
 	mouse = 0;
 	mouses = 0;
@@ -120,7 +120,8 @@
 	case 'p':
 	case 'P':
 		{
-			printoutscreen(number,-1);
+			//printoutscreen(number,-1);
+			printoutscreen(number, -1, view_type, diffusion_number);
 		}
 		break;
 	case 'W':
@@ -346,13 +347,23 @@
 			{
 				if (number%1==0)
 				{
+					// Output cell file at every step
+					cell_file_format();
+					// Output images
 					int vtype = view_type;
 					int dnumber = diffusion_number;
 					bool stext = show_text;
 					view_type = 0;
-					printoutscreen(number,-1);
-					view_type = 2;show_text = true;
+					glClearColor(1.0, 1.0, 1.0, 1.0); // set background color to white
+					glClear(GL_COLOR_BUFFER_BIT);
+					Dataprojection();
+					glutSwapBuffers();
+					//printoutscreen(number,-1);
+					printoutscreen(number, -1, 0, 0);
+					view_type = 2;
+					show_text = true;
 					int dn = reader->get_biology()->getNumRealSpecies();
+					//cout << "    ##### SPECIES: " << dn << endl;
 					for (int i=0;i<dn;i++)
 					{
 						diffusion_number = i;
@@ -360,7 +371,8 @@
 						glClear(GL_COLOR_BUFFER_BIT);
 						Dataprojection();
 						glutSwapBuffers();
-						printoutscreen(number,-1);
+						//printoutscreen(number,-1);
+						printoutscreen(number, -1, 2, i);
 					}
 					view_type = vtype;
 					diffusion_number = dnumber;
@@ -369,11 +381,12 @@
 			}
 			else
 			{
-				printoutscreen(number,error_id);
+				//printoutscreen(number,error_id);
+				printoutscreen(number, error_id, view_type, diffusion_number);
 				cout<<"  -> Bug detected! Please check out cell "<<error_id<<" at output"<<number<<".error.bmp!"<<endl;
 				exit(0);
 			}
-			if (number>=200) pause = true;
+			if (number>=400) pause = true;
 			//pause = true;
 		}
 	}
@@ -489,20 +502,31 @@
 	}
 }
 
-void cellproject::printoutscreen(int number, int id)
+void cellproject::printoutscreen(int number, int id, int viewtype, int diffnum)
 {
+	//cout << "     #### VT -- DN -- NUM: " << viewtype << " -- " << diffnum << " -- " << number;
 	ostringstream temp;
 	temp<<number;
 	string tmps = temp.str();
 	string cwd("");
+	string outname = "";
+	string Molecule = "";
 	// Get the current working directory: 
 	cwd.assign(GetCurrentDir( NULL, 0 ));
-	string outname = cwd+"/out/output"+tmps+".bmp";
-	if (view_type==2 && show_text)
+	if (viewtype == 0)
 	{
-		string Molecule = reader->get_biology()->getSpeciesName(diffusion_number);
-		outname = cwd+"/out/output"+Molecule+tmps+".bmp";
+		outname = cwd+"/out/output_"+tmps+".bmp";
 	}
+	else if (viewtype==2 && show_text)
+	{
+		Molecule = reader->get_biology()->getSpeciesName(diffnum);
+		outname = cwd+"/out/output"+Molecule+"_"+tmps+".bmp";
+	}
+	//cout << "; " << outname << endl;
+	glClearColor(1.0, 1.0, 1.0, 1.0); // set background color to white
+	glClear(GL_COLOR_BUFFER_BIT);
+	Dataprojection();
+	glutSwapBuffers();
 	if (id>=0)
 	{
 		outname = cwd+"/out/output"+tmps+".error.bmp";
--- cellproject.h (trunk r452)
+++ cellproject.h (working copy)
@@ -99,7 +99,8 @@
 		void drawforcedegree(double M[3][3]);
 		void drawNumber(point2D *position, int number, double color[3]);
 		void drawText(point2D *position, string *Text_str, double color[3]);
-		void printoutscreen(int number, int id);
+		//void printoutscreen(int number, int id);
+		void printoutscreen(int number, int id, int viewtype, int diffnum);
 		void drawBresenhamCircle(point2D *center,int radius, double color[3]);
 		void Dataprojection();
 };
--- dbReader.cpp (trunk r485)
+++ dbReader.cpp (working copy)
@@ -232,7 +232,7 @@
 				int dead = atoi(array1[1].c_str());
 				int set_dead = atoi(array1[2].c_str());
 				int migrate = atoi(array1[3].c_str());
-				int soften = atoi(array1[4].c_str());
+				int mAngle = atoi(array1[4].c_str());
 				int type = atoi(array1[5].c_str());
 				double iarea = atof(array1[8].c_str());
 				int nspecies = atoi(array1[9].c_str()); // YFC: READ STATE VECTOR LENGTH
@@ -244,8 +244,8 @@
 				else if (set_dead==1) {cellList[cid]->set_set_dead(true);}
 				if      (migrate==0) {cellList[cid]->set_migrate(false);}
 				else if (migrate==1) {cellList[cid]->set_migrate(true);}
-				if      (soften==0) {cellList[cid]->set_soften(false);}
-				else if (soften==1) {cellList[cid]->set_soften(true);}	
+				//if      (soften==0) {cellList[cid]->set_soften(false);}
+				//else if (soften==1) {cellList[cid]->set_soften(true);}
 				cellList[cid]->set_cell_type(type);
 				cellList[cid]->set_r(CR);
 				cellList[cid]->set_life(number);
@@ -253,6 +253,7 @@
 				cellList[cid]->setup_tcoef(tension_edge_class[type]);
 				cellList[cid]->setup_pcoef(pressure_edge_class[type]);
 				cellList[cid]->setup_mcoef(mass_edge_class[type]);
+                cellList[cid]->set_center(atof(array1[6].c_str()), atof(array1[7].c_str()));
 				// YFC: READ CELL STATE
 				for (int j=0; j<nspecies; j++)
 				{
@@ -888,7 +889,7 @@
 	string tmps = temp.str();
 	string cname = "printout"+tmps+".txt";
 	ofstream OUT (cname.c_str());
-	OUT<<"#######################\n# Cell Growth Project #\n# Sample Database     #\n#######################"<<endl;
+	/**/OUT<<"#######################\n# Cell Growth Project #\n# Sample Database     #\n#######################"<<endl;
 	OUT<<"Point-based\n"<<endl;
 	OUT<<"# This is the bounds of your viewing window plane in VRC coordinates"<<endl;
 	OUT<<"u -100.0,100.0\nv -75.0,75.0"<<endl;
@@ -921,19 +922,21 @@
 		OUT<<index<<","<<p1_id<<","<<p2_id<<endl;
 	}
 	OUT<<"END_EDGE_LIST"<<endl;
+	OUT<<"#ID,Dead,S_Dead,Mig,mAng,Type,X,Y,Area,N_Species,Species0,Species1,..."<<endl;
 	OUT<<"CELL_LIST"<<endl;
 	for (int i=0;i<cn;i++)
 	{
 		int index = cellList[i]->id();
 		int dead = cellList[i]->Dead();
 		int set_dead = cellList[i]->Set_Dead();
-		int migrate = cellList[i]->Migrate();
-		int soften = cellList[i]->Soften();
+		int migrate = cellList[i]->Mark_number(); // 0: static; 1: migrating
+		//int soften = cellList[i]->Soften();
+		double mAngle = cellList[i]->get_migrate_angle();
 		int type = cellList[i]->get_cell_type();
 		double cx = cellList[i]->get_center(0);
 		double cy = cellList[i]->get_center(1);
 		double iarea = cellList[i]->get_area();
-		OUT<<index<<","<<dead<<","<<set_dead<<","<<migrate<<","<<soften<<","<<type<<","<<cx<<","<<cy<<","<<iarea;
+		OUT<<index<<","<<dead<<","<<set_dead<<","<<migrate<<","<<mAngle/*soften*/<<","<<type<<","<<cx<<","<<cy<<","<<iarea;
 		// YFC: ADD OUTPUT FOR CELL STATE
 		int maxlen = max(cellList[i]->get_cell_statelen(), biology->getNumSpecies());
 		int len = cellList[i]->get_cell_statelen();
@@ -949,11 +952,11 @@
 				OUT<<",0";
 			}
 		}
-		for (int j=0;j<(int)cellList[i]->get_sides()->size();j++)
+		/*for (int j=0;j<(int)cellList[i]->get_sides()->size();j++)
 		{
 			int eid = cellList[i]->get_sides()->at(j)->id();
 			OUT<<","<<eid;
-		}
+		}*/
 		OUT<<endl;
 	}
 	OUT<<"END_CELL_LIST"<<endl;
@@ -984,7 +987,7 @@
 		OUT<<endl;
 	}
 	OUT<<"END_PAIR_LIST"<<endl;
-	OUT<<"NODE_LIST"<<endl;
+	/**/OUT<<"NODE_LIST"<<endl;
 	for (int i=0;i<nn;i++)
 	{
 		int index = nodeList[i]->id();
@@ -1066,4 +1069,4 @@
 	cout<<"Print out data file done: "<<(t_e_c-t_b_c)*0.001<<" seconds!"<<endl;
 }
 
-#endif
\ No newline at end of file
+#endif
--- mcommon.xml (trunk r454)
+++ mcommon.xml (working copy)
@@ -26,7 +26,7 @@
           <notes> Clot does not diffuse. </notes>
       <parameter id="D_collagen" value="0.0" unit="cm^2 / sec"/>
           <notes> Collagen does not diffuse. </notes>
-      <parameter id="D_PDGF" value="2.78E-8" unit="cm^2 / sec"/>
+      <parameter id="D_PDGF" value="2.78E-9" unit="cm^2 / sec"/>
           <notes>
             Haugh, J, Biophysical Journal 90:2297-2308, 2006. The diffusion of PDGF in the dermis is
             approximately one thirtieth of its value in solution. 2.78E-8 seems too large, so /5: 5.56E-9
@@ -38,7 +38,7 @@
           <notes>
             Thorne, R. J Neurophysiol 92: 3471-3481, 2004.
           </notes>
-      <parameter id="D_TGFb1" value="2.94E-8" unit="cm^2 / sec"/>
+      <parameter id="D_TGFb1" value="2.94E-9" unit="cm^2 / sec"/>
       <parameter id="D_TGFb3" value="2.94E-9" unit="cm^2 / sec"/>
           <notes>
             Murphy et al. Bull Math Biol (2012) 74:1143–1170. For both TGF b1 and b3.
@@ -51,7 +51,8 @@
       <parameter id="D_KGF" value="9.5E-9" unit="cm^2 / sec"/>
           <notes> Not yet justified. </notes>
       <parameter id="D_Procollagen" value="1.0E-10" unit="cm^2 / sec"/>
-      <parameter id="DeltaT" value="1.0" unit="min"/>
+      <parameter id="DeltaT" value="60.0" unit="min"/>
+      <notes> It was 60.0 min. Need to justify. </notes>
     </listOfParameters>
   </model>
 </sbml>
--- mecm.xml (trunk r500)
+++ mecm.xml (working copy)
@@ -41,7 +41,7 @@
       <parameter id="kd_collagen" value="0" constant="false"/>
       <parameter id="kd0_collagen" value="1.0E-9"/>
       <parameter id="kd1_collagen_mmp" value="1.0E-6"/>
-      <parameter id="kd_pdgf" value="1.668e-3" unit="min^-1"/>
+      <parameter id="kd_pdgf" value="1.667e-3" unit="min^-1"/>
           <notes> Menon et al. 2012.  </notes>
       <parameter id="kd_il1" value="1.152E-2" unit="min^-1"/>
           <notes> Menon et al. 2012.  </notes>
@@ -59,7 +59,7 @@
       <parameter id="Prob_Migration" value="0.0"/>
       <parameter id="Prob_Apoptosis" value="0.0" constant="false"/>
       <parameter id="eapoptosis0" value="1"/>
-      <parameter id="eapoptosis_1" value="0.5"/>
+      <parameter id="eapoptosis_1" value="0.005"/>
       <parameter id="eapoptosis_2" value=""/>
     </listOfParameters>
     <listOfRules>
--- mfibroblast.xml (trunk r454)
+++ mfibroblast.xml (working copy)
@@ -51,8 +51,8 @@
       <parameter id="kd_collagen" value="1" constant="false"/>
       <parameter id="kd0_collagen" value="1.0E-9"/>
       <parameter id="kd1_collagen_mmp" value="1.0E-5"/>
-      <parameter id="kd_pdgf" value="1.668e-3" unit="min^-1"/>
-      <notes> Menon et al. 2012.  </notes>
+      <parameter id="kd_pdgf" value="1.668e-2" unit="min^-1"/>
+      <notes> Haugh et al. BJ 2006; Menon et al. 2012.  </notes>
       <parameter id="kd_il1" value="1.152E-2" unit="min^-1"/>
       <notes> Menon et al. 2012.  </notes>
       <parameter id="kd_tgfa" value="1.0E-3"/>
@@ -78,9 +78,9 @@
       <parameter id="fmigration_1" value="0.5"/>
       <parameter id="fmigration_2" value="1"/>
       <parameter id="fmigration_3" value="0.1"/>
-      <parameter id="fapoptosis0" value="1"/>
+      <parameter id="fapoptosis0" value="0.01"/>
       <parameter id="fapoptosis_1" value="0.005"/>
-      <parameter id="fapoptosis_2" value="1"/>
+      <parameter id="fapoptosis_2" value="0.01"/>
       <parameter id="fapoptosis_3" value="1"/>
     </listOfParameters>
     <listOfRules>
--- mkeratinocyte.xml (trunk r454)
+++ mkeratinocyte.xml (working copy)
@@ -26,11 +26,11 @@
       <parameter id="ks_il1" value="1.0"/>
       <parameter id="ks_tgfa" value="0"/>
       <parameter id="ks_egf" value="0"/>
-      <parameter id="ks_tgfb1" value="1.0" constant="false"/>
-      <parameter id="ks0_tgfb1" value="1.0"/>
+      <parameter id="ks_tgfb1" value="0.0"/>
+      <parameter id="ks0_tgfb1" value="0.0"/>
       <parameter id="ks1_tgfb1_kgf" value="1"/>
       <parameter id="ks2_tgfb1_mmp" value="1"/>
-      <parameter id="ks_tgfb3" value="1.0"/>
+      <parameter id="ks_tgfb3" value="0.0"/>
       <parameter id="ks_fgf2" value="1.0"/>
       <parameter id="ks_mmp" value="0" constant="false"/>
       <parameter id="ks0_mmp" value="1.0"/>
--- mmacrophage.xml (trunk r496)
+++ mmacrophage.xml (working copy)
@@ -22,7 +22,9 @@
     <listOfParameters>
       <parameter id="ks_il1" value="0"/>
       <parameter id="ks_egf" value="0.0"/>
-      <parameter id="ks0_tgfb1" value="1.0"/>
+      <parameter id="ks_pdgf" value="8.0"/>
+      <parameter id="ks1_pdgf" value="0.5"/>
+      <parameter id="ks0_tgfb1" value="0.01"/>
       <parameter id="ks_tgfb1_pdgf" value="1.0"/>
       <parameter id="ks_tgfb3" value="0.0"/>
       <parameter id="ks_fgf2" value="0.0"/>
@@ -41,124 +43,68 @@
       <parameter id="kd_mmp" value="1.0E-3"/>
       <parameter id="kd_tnfa" value="0"/>
       <parameter id="kd_kgf" value="1.0E-3"/>
-      <parameter id="Prob_DoNothing" value="0.0" constant="false"/>
-      <parameter id="Prob_GrowthDiv" value="0.0" constant="false"/>
-      <parameter id="Prob_Migration" value="0.0" constant="false"/>
-      <parameter id="Prob_Apoptosis" value="0.0" constant="false"/>
-      <parameter id="fdonothing0" value="1"/>
-      <parameter id="fdonothing_1" value="5.0"/>
-      <parameter id="fgrowthdiv0" value="1"/>
-      <parameter id="fgrowthdiv_1" value="0.05"/>
-      <parameter id="fgrowthdiv_2" value="1"/>
-      <parameter id="fgrowthdiv_3" value="1"/>
-      <parameter id="fmigration0" value="1"/>
-      <parameter id="fmigration_1" value="0.5"/>
-      <parameter id="fmigration_2" value="1"/>
-      <parameter id="fmigration_3" value="0.1"/>
-      <parameter id="fapoptosis0" value="1"/>
-      <parameter id="fapoptosis_1" value="0.005"/>
-      <parameter id="fapoptosis_2" value="1"/>
-      <parameter id="fapoptosis_3" value="1"/>
+      <parameter id="Prob_DoNothing" value="1.0"/>
+      <parameter id="Prob_GrowthDiv" value="0.0"/>
+      <parameter id="Prob_Migration" value="0.0"/>
+      <parameter id="Prob_Apoptosis" value="0.01"/>
+      <parameter id="mapoptosis0" value="0.01"/>
+      <parameter id="mapoptosis_1" value="0.005"/>
+      <parameter id="mapoptosis_2" value="1"/>
+      <parameter id="mapoptosis_3" value="1"/>
     </listOfParameters>
     <listOfRules>
-      <assignmentRule variable="Prob_DoNothing">
+      <assignmentRule variable="Prob_Apoptosis">
         <math xmlns="http://www.w3.org/1998/Math/MathML">
-          <apply>
-            <times/>
-            <ci> fdonothing0 </ci>
-            <ci> fdonothing_1 </ci>
-          </apply>
-        </math>
-      </assignmentRule>
-      <assignmentRule variable="Prob_GrowthDiv">
-        <math xmlns="http://www.w3.org/1998/Math/MathML">
-          <apply>
-            <divide/>
             <apply>
               <times/>
-              <ci> fgrowthdiv0 </ci>
-              <ci> fgrowthdiv_1 </ci>
+              <ci> mapoptosis0 </ci>
+              <ci> mapoptosis_1 </ci>
               <apply>
                 <plus/>
                 <cn> 1 </cn>
                 <apply>
                   <times/>
-                  <ci> fgrowthdiv_2 </ci>
-                  <ci> PDGF </ci>
+                  <ci> mapoptosis_2 </ci>
+                  <ci> FGF2 </ci>
                 </apply>
               </apply>
-            </apply>
-            <apply>
-              <plus/>
-              <cn> 1 </cn>
-              <apply>
-                <times/>
-                <ci> fgrowthdiv_3 </ci>
-                <ci> TGFb1 </ci>
-              </apply>
-            </apply>
-          </apply>
-        </math>
-      </assignmentRule>
-      <assignmentRule variable="Prob_Migration">
-        <math xmlns="http://www.w3.org/1998/Math/MathML">
-          <apply>
-            <divide/>
-            <apply>
-              <times/>
-              <ci> fmigration0 </ci>
-              <ci> fmigration_1 </ci>
               <apply>
                 <plus/>
                 <cn> 1 </cn>
                 <apply>
                   <times/>
-                  <ci> fmigration_2 </ci>
-                  <ci> PDGF </ci>
+                  <ci> mapoptosis_3 </ci>
+                  <ci> TGFb1 </ci>
                 </apply>
               </apply>
             </apply>
-            <apply>
-              <plus/>
-              <cn> 1 </cn>
-              <apply>
-                <times/>
-                <ci> fmigration_3 </ci>
-                <ci> TGFb1 </ci>
-              </apply>
-            </apply>
-          </apply>
         </math>
       </assignmentRule>
-      <assignmentRule variable="Prob_Apoptosis">
-        <math xmlns="http://www.w3.org/1998/Math/MathML">
+    </listOfRules>
+    <listOfReactions>
+      <reaction id="SynPDGF" reversible="false" fast="false">
+        <listOfProducts>
+          <speciesReference species="PDGF"/>
+        </listOfProducts>
+        <kineticLaw>
+          <math xmlns="http://www.w3.org/1998/Math/MathML">
             <apply>
               <times/>
-              <ci> fapoptosis0 </ci>
-              <ci> fapoptosis_1 </ci>
+              <ci> default </ci>
+              <ci> ks_pdgf </ci>
               <apply>
                 <plus/>
-                <cn> 1 </cn>
+                <cn> 1.0 </cn>
                 <apply>
                   <times/>
-                  <ci> fapoptosis_2 </ci>
-                  <ci> FGF2 </ci>
+                  <ci> ks1_pdgf </ci>
+                  <ci> PDGF </ci>
                 </apply>
               </apply>
-              <apply>
-                <plus/>
-                <cn> 1 </cn>
-                <apply>
-                  <times/>
-                  <ci> fapoptosis_3 </ci>
-                  <ci> TGFb1 </ci>
-                </apply>
-              </apply>
             </apply>
-        </math>
-      </assignmentRule>
-    </listOfRules>
-    <listOfReactions>
+          </math>
+        </kineticLaw>
+      </reaction>
       <reaction id="SynTGFb1" reversible="false" fast="false">
         <listOfProducts>
           <speciesReference species="TGFb1"/>
--- para.h (trunk r490)
+++ para.h (working copy)
@@ -28,6 +28,7 @@
 const double TM = 0.1;           // time step
 const double BD = 2*PI*CR/SN;    // edge size of initialized polygon
 const double AR = 2*CR*CR*2.598; // double size of the hexagon inside the shpere
+const double AR0 = CR*CR*2.598;	 // resting cell area
 const double GAR = PI*CR*CR/10;  // incremental area
 const double RDT = 24.0;         // cell cycle time: 24 hours
 //const double MU = 1.91;
--- point3D.cpp (trunk r488)
+++ point3D.cpp (working copy)
@@ -749,7 +749,9 @@
 	set_dead = false;
 	migrate = false;
 	migrate_relax = false;
+	migrate_angle = -1;
 	mark_number = 0; // default
+	mAngle = 0;
 	soften = false;
 	migrate_response = false;
 	circle = true;
--- point3D.h (trunk r507)
+++ point3D.h (working copy)
@@ -314,6 +314,7 @@
 		double sense_migr_vec[2]; // mechanical signal strength and orientation in response with migration
 		double sense_migr_angle; // mechanical signal direction: keep for 3 steps
 		int mark_number; // 0: normal; 1: migrate; 2:soften
+		int mAngle;  // migration angle
 		bool soften;     // the cell is soften
 		bool migrate_response; // the cells on the counter direction of migration
 		int cell_status; // -1: shrink; 0: static; 1: grow
--- util.cpp (trunk r512)
+++ util.cpp (working copy)
@@ -389,7 +389,7 @@
 		{
 			w[ia->at(i)] += a->at(i) * v->at(ja->at(i)) * dt;
 		}
-		nnegv = 0;
+		/**/nnegv = 0;
 		for (int i = 0; i < n; i++)
 		{
 			testv = v->at(i) + w[i];
@@ -408,7 +408,7 @@
 				err = max (err, w[i]);
 			}
 		}
-		else if (dt > 1e-4)
+		/**/else if (dt > 1e-4)
 		{
 			dt = dt/10.0;
 			cout << "          WARNING: Concentration < 0, big time step suspected, change to smaller dt=" << dt << endl;
@@ -423,8 +423,150 @@
 		ns ++;
 	}
 	delete[] w;
+	return err;
+}
+
+double RK4 (vector<int>* ia, vector<int>* ja, vector<double>* a, vector<double>* v, double t)
+{
+	/***********************************************************
+	  Linear Runge-Kutta solver for finite element diffusion equation
+	  dx/dt = Ax
+	  YOUFANG CAO
+	***********************************************************/
+	double dt = 0.1;
+	double dthalf = 0.5 * dt;
+	double h6 = dt / 6.0;
+	double err = 0.0;
+	double tau = 0;
+	double testv = 0;
+	int    nnegv = 0;
+	int    n = v->size();
+	int    nnz = ia->size ();
+	int    ns = 0;
+	double maxv = 0;
+	double maxa = 0;
+	for (int i = 0; i < n; i++)
+	{
+		if (v->at(i) > maxv) { maxv = v->at(i); }
+	}
+	if (maxv < 1e-4) { return 0.0; }
+	for (int i = 0; i < nnz; i++)
+	{
+		if (abs(a->at(i)) > maxa) { maxa = abs(a->at(i)); }
+	}
+	if (maxa < 1e-8) { return 0.0; }
+
+
+	double *w1;
+	double *w2;
+	double *w3;
+	double *w4;
+	double *w0;
+	double *delta;
+
+	w1 =  new double[n];
+    w2 =  new double[n];
+    w3 =  new double[n];
+    w4 =  new double[n];
+    w0 =  new double[n];
+	delta =  new double[n];
+	
+	while(tau < t)
+	{
+		for (int i=0; i<n; i++) 
+		{
+			w0[i] = w1[i] = w2[i] = w3[i] = w4[i] = 0.0;
+			delta[i] = 0.0;
+		}
+
+		for (int i=0; i<nnz; i++)
+		{
+			w1[ia->at(i)] += a->at(i) * v->at(ja->at(i)); 
+		}
+
+		for (int i=0; i<n; i++)
+		{
+			w0[i] = v->at(i) + dthalf * w1[i];
+		}
+
+		for (int i=0; i<nnz; i++)
+		{
+			w2[ia->at(i)] += a->at(i) * w0[ja->at(i)];
+		}
+
+		for (int i=0; i<n; i++)
+		{
+			w0[i] = v->at(i) + dthalf * w2[i];
+		}
+
+		for (int i=0; i<nnz; i++)
+		{
+			w3[ia->at(i)] += a->at(i) * w0[ja->at(i)];
+		}
+
+		for (int i=0; i<n; i++)
+		{
+			w0[i] = v->at(i) + dt * w3[i];
+		}
+
+		for (int i=0; i<nnz; i++)
+		{
+			w4[ia->at(i)] += a->at(i) * w0[ja->at(i)];
+		}
+
+		nnegv = 0;
+		err = 0;
+		for (int i = 0; i < n; i++)
+		{
+			delta[i] = h6 * (w1[i] + 2*w2[i] + 2*w3[i] + w4[i]);
+			testv = v->at(i) + delta[i];
+			if (testv < 0)
+			{
+				nnegv ++;
+				break;
+			}
+			if (abs(delta[i]) > err)
+			{
+				err = abs(delta[i]);
+			}
+		}
+
+		if (nnegv == 0)
+		{
+			for (int i = 0; i < n; i++)
+			{
+				v->at(i) += delta[i];
+			}
+		}
+		else if (nnegv > 0)
+		{
+			if (dt > 1e-4)
+			{
+				dt = dt/10.0;
+				cout << "          WARNING: Concentration < 0, big time step suspected, change to smaller dt=" << dt << endl;
+				continue;
+			}
+			else
+			{
+				cout << "          WARNING: dt too small, RK4 not done." << endl;
+				break;
+			}
+		}
+
+		tau += dt;
+	}
+
+	delete[] w0;
+	delete[] w1;
+	delete[] w2;
+	delete[] w3;
+	delete[] w4;
+
 	return err;
 }
 
+
+
+
 #endif
 
--- util.h (trunk r509)
+++ util.h (working copy)
@@ -41,5 +41,6 @@
 double U_to_N(); // generate normal distribution from uniform distribution
 double SOR_solver(int sm_idt, double converge, double w, int steps, int &flagtime, int *SOR_N, int *JA, double *VA, double *F, double *Xn, double *X, double *SOR_II_V);
 double Euler (vector<int>* ia, vector<int>* ja, vector<double>* a, vector<double>* v, double t);
+double RK4 (vector<int>* ia, vector<int>* ja, vector<double>* a, vector<double>* v, double t);
 
 #endif
\ No newline at end of file
